llama_index/llms/vllm.py

```diff
@@ -0,0 +1,356 @@
+        default=None,
+        description="Number of output sequences that are generated from the prompt.",
+    )
+
+    presence_penalty: float = Field(
+        default=0.0,
+        description="Float that penalizes new tokens based on whether they appear in the generated text so far.",
+    )
+
+    frequency_penalty: float = Field(
+        default=0.0,
+        description="Float that penalizes new tokens based on their frequency in the generated text so far.",
+    )
+
+    top_p: float = Field(
+        default=1.0,
+        description="Float that controls the cumulative probability of the top tokens to consider.",
+    )
+
+    top_k: int = Field(
+        default=-1,
+        description="Integer that controls the number of top tokens to consider.",
+    )
+
+    use_beam_search: bool = Field(
+        default=False, description="Whether to use beam search instead of sampling."
+    )
+
+    stop: Optional[List[str]] = Field(
+        default=None,
+        description="List of strings that stop the generation when they are generated.",
+    )
+
+    ignore_eos: bool = Field(
+        default=False,
+        description="Whether to ignore the EOS token and continue generating tokens after the EOS token is generated.",
+    )
+
+    max_new_tokens: int = Field(
+        default=512,
+        description="Maximum number of tokens to generate per output sequence.",
+    )
+
+    logprobs: Optional[int] = Field(
+        default=None,
+        description="Number of log probabilities to return per output token.",
+    )
+
+    dtype: str = Field(
+        default="auto",
+        description="The data type for the model weights and activations.",
+    )
+
+    download_dir: Optional[str] = Field(
+        default=None,
+        description="Directory to download and load the weights. (Default to the default cache dir of huggingface)",
+    )
+
+    vllm_kwargs: Dict[str, Any] = Field(
+        default_factory=dict,
+        description="Holds any model parameters valid for `vllm.LLM` call not explicitly specified.",
+    )
+
+    api_url: str = Field(description="The api url for vllm server")
+
+    _client: Any = PrivateAttr()
+
+    def __init__(
+        self,
+        model: str = "facebook/opt-125m",
+        temperature: float = 1.0,
+        tensor_parallel_size: Optional[int] = 1,
+        trust_remote_code: Optional[bool] = True,
+        n: int = 1,
+        best_of: Optional[int] = None,
+        presence_penalty: float = 0.0,
+        frequency_penalty: float = 0.0,
+        top_p: float = 1.0,
+        top_k: int = -1,
+        use_beam_search: bool = False,
+        stop: Optional[List[str]] = None,
+        ignore_eos: bool = False,
+        max_new_tokens: int = 512,
+        logprobs: Optional[int] = None,
+        dtype: str = "auto",
+        download_dir: Optional[str] = None,
+        vllm_kwargs: Dict[str, Any] = {},
+        api_url: Optional[str] = "",
+        callback_manager: Optional[CallbackManager] = None,
```

Comment:
We can add a `completion_to_prompt` function hook, so that any calls to `chat()` or `stream_chat()` can be routed to the proper `complete` endpoint. A few LLMs do this, like `llama_cpp`

Reply:
Completed